webui/react/src/components/ManageNodesModal.tsx

```diff
@@ -0,0 +1,93 @@
+import Icon from 'hew/Icon';
+import Input from 'hew/Input';
+import Message from 'hew/Message';
+import { Modal } from 'hew/Modal';
+import Row from 'hew/Row';
+import Toggle from 'hew/Toggle';
+import { Body, Label } from 'hew/Typography';
+import { isEqual } from 'lodash';
+import React, { useCallback, useState } from 'react';
+
+import { disableAgent, enableAgent } from 'services/api';
+import { Agent } from 'types';
+import handleError from 'utils/error';
+
+import css from './ManageNodesModal.module.scss';
+
+interface Props {
+  nodes: Agent[];
+}
+
+// const defaultNodes: Agent[] = [
+//   { id: 'ABC', enabled: true, registeredTime: 1000, resourcePools: [], resources: []},
+//   { id: 'DEF', enabled: false, registeredTime: 2000, resourcePools: [], resources: [] },
+// ];
+
+const ManageNodesModalComponent = ({ nodes }: Props): JSX.Element => {
+  // nodes = defaultNodes;
+  const originalNodes = nodes.reduce((obj: Record<string, boolean>, node: Agent) => {
+    obj[node.id] = !!node.enabled;
+    return obj;
+  }, {});
+
+  const [searchText, setSearchText] = useState<string>('');
+  const [toggleStates, setToggleStates] = useState<Record<string, boolean>>(originalNodes);
+
+  const onFilterChange = useCallback((e: React.ChangeEvent<HTMLInputElement>) => {
+    setSearchText(e.target.value);
+  }, []);
+
+  const onToggleNode = useCallback((nodeId: string) => {
+    setToggleStates((prev) => ({ ...prev, [nodeId]: !prev[nodeId] }));
+  }, []);
+
+  const onFormSubmit = useCallback(async () => {
+    for (const nodeId in toggleStates) {
+      if (toggleStates[nodeId] !== originalNodes[nodeId]) {
```

Comment:
consider memoizing `originalNodes` so each render doesn't recalculate the node state. Alternatively (assuming the endpoints are idempotent), keep track of which endpoints were touched and apply the endpoints that way.

Reply:
✔️ added `useMemo`